Daryun

dksekfbs72 opened the pull request:
# PR 생성 날짜
- 2022.12.04

# 변경 사항
- 회원 가입 기능 추가
- 관리자 회원 관리 기능 추가
- Entity class 표기법을 스네이크 표기법 -> 카멜 표기법 변경
- security, Junit5, mockito 의존성 추가
- DB명 restaurant -> store 변경 기획 템플릿도 변경 예정

# postman 테스트 결과
- 회원 가입 기능 정상
- 관리자 회원 관리 기능 정상

# 궁금한 사항
- user 테이블의 password를 암호화하여 받기 위해 security를 추가했는데, 문제가 있을지 궁금합니다..!
- 테스트 createAccountSuccess에서 가입한 회원 정보를 userRepository로 불러와서 Assertions로 비교하려 했지만 실제로 데이터가 저장되지 않아서 실패했는데, 어떻게 고치면 가능할지 궁금합니다
- 기타 오류를 핸들링 하는 방법이 감이 잡히지 않아서 조언 부탁드립니다..!

dksekfbs72 committed:
test

dksekfbs72 committed:
UserController/UserRepository/UserInput/UserService 추가 , 회원가입 기능 추가, Entity 스네이크 표기법 -> 카멜 표기법, 회원가입 테스트 추가

회원가입 기능은 추가했으나 오류 탐색은 아직 추가하지 않았음.
Entity 클래스는 스네이크 표기법보다 카멜 표기법이 적합할 것 같아 변경함.
회원가입 시에 회원가입한 정보를 반환함.
테스트가 원시적이라서 리팩토링이 필요함
security, Junit5, dependency 의존성 추가

dksekfbs72 committed:
회원가입 기능 추가, 관리자 회원 관리 기능 추가, Entity 스네이크 표기법 -> 카멜 표기법

회원가입 기능은 추가했으나 오류 탐색은 아직 추가하지 않았음.
Entity 클래스는 스네이크 표기법보다 카멜 표기법이 적합할 것 같아 변경함.
테스트가 원시적이라서 리팩토링이 필요함.
관리자의 회원 정보 변경 기능은 유저의 아이디와 비밀번호는 변경할 수 없음.
security, Junit5, dependency 의존성 추가

dksekfbs72 committed:
회원가입 기능 추가, 관리자 회원 관리 기능 추가, Entity 스네이크 표기법 -> 카멜 표기법

회원가입 기능은 추가했으나 오류 탐색은 아직 추가하지 않았음.
Entity 클래스는 스네이크 표기법보다 카멜 표기법이 적합할 것 같아 변경함.
테스트가 원시적이라서 리팩토링이 필요함.
관리자의 회원 정보 변경 기능은 유저의 아이디와 비밀번호는 변경할 수 없음.
security, Junit5, dependency 의존성 추가

dksekfbs72 committed:
회원가입 기능 추가, 관리자 회원 관리 기능 추가, Entity 스네이크 표기법 -> 카멜 표기법

회원가입 기능은 추가했으나 오류 탐색은 아직 추가하지 않았음.
Entity 클래스는 스네이크 표기법보다 카멜 표기법이 적합할 것 같아 변경함.
테스트가 원시적이라서 리팩토링이 필요함.
관리자의 회원 정보 변경 기능은 유저의 아이디와 비밀번호는 변경할 수 없음.
security, Junit5, dependency 의존성 추가

KimHyoJin commented on src/main/java/zerobase/demo/controller/AdminUserController.java:
```diff
@@ -0,0 +1,30 @@
+	String createUser(@ModelAttribute UserDto parameter,
+		@RequestParam String myId) {
+		//유저의 아이디와 비밀번호는 변경할 수 없음
+		System.out.println(myId);
```
system.out.println 대신 logger를 사용하시는걸 추천드립니다. 아래 블로그를 한번 참고해보시면 어떤 장점이 있는지 확인하실수 있으실것 같아요 https://recordsoflife.tistory.com/281

KimHyoJin commented on src/main/java/zerobase/demo/entity/Menu.java:
```diff
@@ -14,14 +14,15 @@
 	private int price;
 	private String name;
-	private String picture_url;
+	private String pictureUrl;
```
👍🏻

KimHyoJin commented on src/main/java/zerobase/demo/service/UserService.java:
```diff
@@ -0,0 +1,79 @@
+
+	private final UserRepository userRepository;
+
+	public ResponseResult createUser(UserInput parameter) {
```
ResponseResult는 controller에서 내려주는 응답 형태라고 이해했습니다! 이 부분은 service layer까지 오지 않는 것이 좋을 것 같아요. layer별 dto를 명확히 나누면 좋을 것 같습니다. 
대신 이 service layer에서는 exception을 던지고, @controllerAdvice를 통해 exception을 잡아 이를 ResponseResult형태로 내려주면 좋을 것 같아요

KimHyoJin replied:
https://jeong-pro.tistory.com/195

KimHyoJin replied:
마찬가지로 input도 controller에서 사용하는 것을 그대로 받아오는 것보단 각 필드를 풀어서 받아오거나 다른 dto로 받아오는 방법을 추천드립니다!

KimHyoJin replied:
controllerAdvice는 같은 프로젝트 다른 PR에서 구현하신것 같네요! 두분이서 한번 논의해보시면 좋을 것 같습니다 :) https://github.com/dksekfbs72/zogiyo/pull/3/files

KimHyoJin commented on src/main/java/zerobase/demo/model/ResponseResult.java:
```diff
@@ -0,0 +1,15 @@
+@AllArgsConstructor
+@Builder
+public class ResponseResult {
+	boolean result;
```
성공여부를 나타내는 필드인가용? success라는 단어도 괜찮을 것 같습니다.

KimHyoJin commented on src/main/java/zerobase/demo/service/UserService.java:
```diff
@@ -0,0 +1,79 @@
+	}
+
+	public ResponseResult adminUpdateUser(UserDto parameter, String myId) {
+		Optional<User> optionalAdmin = userRepository.findByUserId(myId);
```
`User optionalAdmin = userRepository.findByUserId(myId).orElseThrow(() -> new Exception()); `
형태나(위 형태는 제가 위에서 조언드린 형태(exception을 던지는)에요) 
지금과 같은 코드돠 동일하게 동작하려면 
`User optionalAdmin = userRepository.findByUserId(myId).orElse(new ResponseResult(false,"아이디가 존재하지 않습니다.")); `

를 사용하시면 가독성이 좀더 높아질 것 같습니다

KimHyoJin commented on src/main/java/zerobase/demo/service/UserService.java:
```diff
@@ -0,0 +1,79 @@
+			return new ResponseResult(false,"이미 존재하는 아이디입니다.");
+		}
+
+		if (!parameter.getStatus().equals("user") && !parameter.getStatus().equals("owner")) {
```
status같은 경우 어떤 지정된 몇가지 상태가 있을 것 같은데요, enum으로 관리하시면 좀더 깔끔할 것 같아요

KimHyoJin commented on src/main/java/zerobase/demo/config/AllExceptionHandler.java:
```diff
@@ -0,0 +1,17 @@
+	@ExceptionHandler(Exception.class)
+	//모든 오류를 핸들링
+	public Exception handleAllException() {
+		return new Exception();
```
이쪽부분에서 exception을 다시 던져주는 것이 아닌 형태에 맞는 ResponseResult를 던져주시면 될것 같아요!

KimHyoJin commented on src/main/java/zerobase/demo/controller/UserController.java:
```diff
@@ -0,0 +1,26 @@
+
+		ResponseResult responseResult = userService.createUser(parameter);
+
+		return responseResult.getMessage();
```
요거 왜 ResponseResult로 return하지 않고 getMessage만 return하도록 되어있나용?

KimHyoJin reviewed (approved):
안녕하세요 코드 잘보았습니다! 첫 PR이시지만 매우 깔끔하게 잘 짜셨네요!!! 질문주신거 먼저 답해드릴려고 합니다. 

> user 테이블의 password를 암호화하여 받기 위해 security를 추가했는데, 문제가 있을지 궁금합니다..!

문제 없습니다! 다만 security를 추가하시면서 configuration도 추가하신걸로 보이는데요, 현재는 전체 path에 대해 permitAll이 되어있지만 이 부분을 추후에는 변경해서 전체적으로 프로젝트에 spring-security를 적용시키시는걸 추천드려요 

> 테스트 createAccountSuccess에서 가입한 회원 정보를 userRepository로 불러와서 Assertions로 비교하려 했지만 실제로 데이터가 저장되지 않아서 실패했는데, 어떻게 고치면 가능할지 궁금합니다

unit test를 작성하시려고 하신것처럼 보여요! unit test는 말그대로 지정된 유닛단위의 동작을 테스트하는거라, 테스트 범위 밖의 것들은 mocking을 해주셔야합니다. 
`userRepository`에 대한 mocking을 해주셔야할 것 같아요. 현재 테스트하려는 코드가 createUser라고 이해했습니다! 그러면 `createUser`에서 `userRepository.findByUserId`라는 함수를 하나 호출하고, `userRepository.save` 함수도 호출이 됩니다. 이 두 함수가 호출되었을때 어떻게 동작할 것인지를 when, then을 이용해서 작성하셔야합니다. 
성공할경우는 findByUserId시 empty가 나와야겠죠 그럼 아래처럼 mocking해볼수 있을것같아요
`when(userRepository.findByUserId("Id")).thenReturn(Optional.empty());` 
`userRepository.save`도 이와 유사하게 mocking하시면 되구요, 혹시 user가 존재할경우에 대한 테스트도 이처럼 작성하시면됩니다!

> 기타 오류를 핸들링 하는 방법이 감이 잡히지 않아서 조언 부탁드립니다..! 

exception별로 쪼개서 받으시면 됩니다! 코멘트로 달아둔 블로그를 한번 확인해보시구 잘 이해가 안가시는 부분있으면 슬랙으로 질문주세요 :) 


수고 많으셨습니다!

dksekfbs72 committed:
#exception 작업, logger 작업, UserStatus enum 작업, ResponseResult로 controller 작업 클래스 통일

exception 따로 잡아두고 UserInput 풀어서 UserService 매개 변수로 받음.
UserService 까지는 ResponseResult 를 불러오지 않게 코딩.

dksekfbs72 committed:
로그인 구현 및 관리자 권한 적용

adminUpdateUser 함수에서 관리자 권한을 확인하던 조건문 제거, 대신 Security가 ROLL_ADMIN을 확인, 로그인 성공 페이지 controller에 구현

dksekfbs72 committed:
추상클래스 BaseEntity 추가, 내정보 보기 기능 추가

추상 클래스 BaseEntity를 모든 Entity가 상속받습니다. 내정보 보기 기능 추가했습니다.
